src/rust/engine/graph/README.md

```diff
@@ -0,0 +1,49 @@
+
+Confusingly, "root" in the context of `invalidate_from_roots` means a root in the dependee graph: calling the method for a root `Node` will "clear" the value of that `Node`, and mark the transitive dependees of that `Node` "dirty". No `Node`s are ever deleted from the `Graph`, so "clearing" a `Node` means moving its `Entry` to the `NotStarted` state. Because it will be in the `NotStarted` state, the cleared `Node` will definitely re-run when it is next requested, but any dirty `Node`s that might depend on it will only re-run if they cannot be "cleaned": see below.
+
+### Dirtying and Cleaning
+
+When the `Entry` for a `Node` is marked dirty by "Invalidation" (see above) we only _suspect_ that it might need to re-run, because one of its dependencies has been cleared. To determine whether it _actually_ needs to re-run, the next time the dirty `Node` is requested, `Graph::get` (via [`Entry::get_node_result`](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/entry.rs#L361-L374)) will compare the [`Generation` values](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/entry.rs#L37-L57) of the dependencies used to compute the previous value of the `Node` with their current `Generation` values. If none of the (`Generation` values) of the `Node` have changed, then it does not need to re-run: this is called "cleaning".
+
+When a `Node` _does_ need to re-run for some reason (either due to having been cleared by invalidation, or being marked dirty and then failing to be cleaned), its previous and new result are compared to determine whether its `Generation` value should increment. The `Generation` value avoids the need to keep multiple copies of a `Node`'s result value in memory over time, and makes cleaning cheaper by converting potentially expensive equality checks into integer comparisons.
+
+### Uncacheable Nodes
+
+`Node`s can be marked "uncacheable" ([via](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/node.rs#L32-L35)), which causes them to re-run once per "session" (controlled by a [SessionId](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/node.rs#L97-L101) on the Context), regardless of whether any of their inputs have changed.
```

Comment:
This mean that Pantsd still caches it? It might be helpful to define a session, beyond linking to the code. E.g.

> which is a single Pants run if Pantsd is not used, and is otherwise until Pantsd restarts.

Reply:
It's always a single run, pantsd or no.